### PR TITLE
ARROW-1716: [Format/JSON] Use string integer value for Decimals in JSON

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java [hunk 1]
```diff
@@ -332,9 +334,10 @@ private void setValueFromParser(ValueVector valueVector, int i) throws IOExcepti
         ((Float8Vector) valueVector).getMutator().set(i, parser.readValueAs(Double.class));
         break;
       case DECIMAL: {
-          DecimalVector decimalVector = ((DecimalVector) valueVector);
-          byte[] value = decodeHexSafe(parser.readValueAs(String.class));
-          DecimalUtility.writeByteArrayToArrowBuf(value, decimalVector.getBuffer(), i);
+          DecimalVector decimalVector = (DecimalVector) valueVector;
+          BigInteger integerValue = new BigInteger(parser.readValueAs(String.class));
+          BigDecimal decimalValue = new BigDecimal(integerValue, 0);
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Could you add a note that the scale won't be used when writing to the ArrowBuf?  Also, I think you could just use `BigDecimal(integerValue)` to make an unscaled value

**Review by BryanCutler — COMMENTED** [review]
Looks good, but this should fail integration tests now until we swap bytes from LE for Java right?

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java (hunk 1 above)
Comment (cpcloud) [reply java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Yep will do.

**Comment by cpcloud** [comment]
@BryanCutler Yep, that's right. The integration tests are currently failing. Can you help me out with making the changes to the Java codebase required to get the integration tests working again? I'm happy to keep digging, but this will go much faster if someone familiar with the Java side of things can give me a hand.

**Comment by BryanCutler** [comment]
Sure, I can look into it tomorrow

**Comment by icexelloss** [comment]
@cpcloud I think either @BryanCutler or I can work with you tomorrow. @BryanCutler feel free to take a look, if you need me to take this let me know.

**Comment by cpcloud** [comment]
@BryanCutler @icexelloss Thanks guys.

**Comment by BryanCutler** [comment]
@cpcloud I create https://github.com/cpcloud/arrow/pull/1
Not sure if it's the most optimized way to swap bytes, but integration tests are passing for me

File: cpp/src/arrow/ipc/json-internal.cc [hunk 2]
```diff
@@ -1094,6 +1108,32 @@ class ArrayReader {
...
+
+    for (int i = 0; i < length_; ++i) {
+      if (!is_valid_[i]) {
+        RETURN_NOT_OK(builder.AppendNull());
+        continue;
+      }
+
+      const rj::Value& val = json_data_arr[i];
+      DCHECK(val.IsString()) << "JSON value is not a string";
+      const Decimal128 value(val.GetString());
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Can this fail?

File: cpp/src/arrow/test-util.h [hunk 3]
```diff
@@ -143,55 +139,161 @@ static inline Status GetBitmapFromVector(const std::vector<T>& is_valid,
 // Sets approximately pct_null of the first n bytes in null_bytes to zero
 // and the rest to non-zero (true) values.
 static inline void random_null_bytes(int64_t n, double pct_null, uint8_t* null_bytes) {
-  Random rng(random_seed());
-  for (int64_t i = 0; i < n; ++i) {
-    null_bytes[i] = rng.NextDoubleFraction() > pct_null;
-  }
+  const int random_seed = 0;
+  std::mt19937 gen(random_seed);
```
Comment (wesm) [inline cpp/src/arrow/test-util.h]: Use a random seed here? Or add a seed option to the function

File: cpp/src/arrow/test-util.h [hunk 4]
```diff
@@ -143,55 +139,161 @@ static inline Status GetBitmapFromVector(const std::vector<T>& is_valid,
...
 }
 
 static inline void random_is_valid(int64_t n, double pct_null,
                                    std::vector<bool>* is_valid) {
-  Random rng(random_seed());
-  for (int64_t i = 0; i < n; ++i) {
-    is_valid->push_back(rng.NextDoubleFraction() > pct_null);
-  }
+  const int random_seed = 0;
+  std::mt19937 gen(random_seed);
```
Comment (wesm) [inline cpp/src/arrow/test-util.h]: Seed

File: cpp/src/arrow/test-util.h [hunk 5]
```diff
@@ -81,25 +83,20 @@ namespace test {
...
-  uint64_t draw;
-  uint64_t span = upper - lower;
-  T val;
-  for (int64_t i = 0; i < N; ++i) {
-    draw = rng.Uniform64(span);
-    val = static_cast<T>(draw + lower);
-    out->push_back(val);
-  }
+  const int random_seed = 0;
+  std::mt19937 gen(random_seed);
```
Comment (wesm) [inline cpp/src/arrow/test-util.h]: Random or static seed?

Comment (cpcloud) [reply cpp/src/arrow/test-util.h]: This was originally calling the `random_seed()` function which just returned zero. I can add a seed parameter.

**Comment by cpcloud** [comment]
@wesm Would you mind if I address the seed parameter in a follow up PR? There's already a bunch of changes in this PR and I'd rather deal with any change in test behavior after this is merged in.

**Comment by wesm** [comment]
Sure, no worries about the seed

File: cpp/src/arrow/ipc/json-internal.cc (hunk 2 above)
Comment (cpcloud) [reply cpp/src/arrow/ipc/json-internal.cc]: I'll use the `Decimal::FromString` static method here since that'll return `Status`.

**Comment by cpcloud** [comment]
Good news is that we should be able to print the full range of values that fit into 16 bytes, including those that require more than 38 decimal digits to be displayed.

**Comment by wesm** [comment]
Rebase this? Then this should be good

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**